kpops/pipeline_generator/pipeline.py

```diff
@@ -69,6 +80,53 @@ def __iter__(self) -> Iterator[PipelineComponent]:
     def __len__(self) -> int:
         return len(self.components)
 
+    def build_execution_graph_from(
+        self,
+        components: list[PipelineComponent],
+        reverse: bool,
+        runner: Callable[[PipelineComponent], Coroutine],
+    ):
+
+        async def run_parallel_tasks(tasks):
+            await asyncio.gather(*tasks)
+
+        async def run_graph_tasks(pending_tasks: list[Awaitable]):
+            for pending_task in pending_tasks:
+                await pending_task
+
+        nodes = [node_component.id for node_component in components]
+        transformed_graph = self.graph.subgraph(nodes)
+
+        if reverse:
+            transformed_graph = transformed_graph.reverse()
+
+        root_node = "root_node_bfs"
+        transformed_graph.add_node(root_node)
+
+        for node in self.graph:
+            predecessors = list(self.graph.predecessors(node))
+            if not predecessors:
+                transformed_graph.add_edge(root_node, node)
+
+        layers_graph = list(nx.bfs_layers(transformed_graph, root_node))
+
+        sorted_tasks = []
+
+        for layer in layers_graph[1:]:
+            parallel_tasks = []
+            for task in layer:
+                self.__get_parallel_task_from(parallel_tasks, runner, task)
+
+            if parallel_tasks:
+                sorted_tasks.append(run_parallel_tasks(parallel_tasks))
+
+        return run_graph_tasks(sorted_tasks)
+
+    def __get_parallel_task_from(self, parallel_tasks, runner, task):
+        node = self._component_index[task]
+        if node.component is not None and not node.is_topic:
+            parallel_tasks.append(runner(node.component))
```

Comment:
Two things here:
1. It is much clearer if you return the `parallel_tasks` list from this method and remove the input list.
2. I don't understand why you made a difference between the topic and the components. 🤔

Reply:
2.) because topics doesn't have something to be run and for preprocessing proposes inside the graph creation, we say that the topic is an extra node.